Add libSyntax support for #gradient, #valueAndGradient, #adjoint.

dan-zheng opened the pull request:
- Define GradientExpr and AdjointExpr syntax nodes.
  - Share code with `@differentiable` attribute node (DifferentiationParameterList).
- Simplify `parseTypeIdentifier` function (reduce custom logic).
- Add round-trip parse tests.

dan-zheng committed:
Add libSyntax support for #gradient, #valueAndGradient, #adjoint.

- Define GradientExpr and AdjointExpr syntax nodes.
  - Share code with @differentiable attribute node (DifferentiationParameterList).
- Simplify `parseTypeIdentifier` function (reduce custom logic).
- Add round-trip parse tests.

dan-zheng commented:
@nkcsgexi Could you please review this PR? That would be appreciated!

dan-zheng commented on utils/gyb_syntax_support/ExprNodes.py:
```diff
@@ -569,4 +569,62 @@
+                   ]),
+             Child('Dot', kind='PeriodToken', is_optional=True,
+                   description='The period after the optional base type.'),
+             Child('DeclBaseName', kind='Syntax', description='''
```
There are other nodes that contain a "DeclName", aka an `Identifier`/`XXOperator` followed by `DeclNameArguments`.
It might be good to add a "DeclName" node upstream and use that. It may be the case that some nodes allow operators while others don't, however.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Merge branch 'tensorflow' into libsyntax-autodiff

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented:
I haven't looked into the implementation, but unifying the param list syntax of `@differentiable` and `#gradient` sounds a bit dangerous. `@differentiable` supports a `self` argument whereas `#gradient` doesn't.

nkcsgexi commented on lib/Parse/ParseType.cpp:
```diff
@@ -646,25 +675,25 @@ Parser::parseTypeIdentifier(bool isParsingQualifiedDeclName) {
-    if (ComponentsR.empty())
+    // If components are empty, discard the syntax parsing context.
+    if (ComponentsR.empty()) {
+      IdentTypeCtxt.setDiscard();
```
Any reason we're using `setDiscard()` here instead of using `setTransparent()`? We usually only use `setDiscard()` in back tracking context. One goal of syntax tress is we can round-trip print even invalid code. Otherwise, LGTM!

dan-zheng replied:
Good point, thanks! I didn't expect `setTransparent` to work in place of `setDiscard`, but it did which is nice. Done in https://github.com/apple/swift/pull/17725/commits/5a64f20b9524636216f575163f60e92537c170fc.

In fact, backtracking is being performed here (I added the `isParsingQualifiedDeclName` argument to `parseTypeIdentifier` to enable backtracking the last type component, [explanation here](https://github.com/apple/swift/pull/17324#issuecomment-398514702)).

dan-zheng commented:
@rxwei It seems `#gradient` currently does [support a `self` argument](https://github.com/apple/swift/blob/e8ee775e791437dfcdd4013fc77b8c4bcbf7d3fa/lib/Parse/ParseExpr.cpp#L3719), at least during parsing. Should `self` not be supported?

dan-zheng committed:
Use `setTransparent` instead of `setDiscard`.

- `setDiscard` is typically used only in backtracking contexts. Backtracking is
  being performed in `parseTypeIdentifier`, but it's better to avoid using
  `setDiscard` if possible.

Addresses comment by @nkcsgexi.

rxwei commented:
It shouldn’t be supported.

dan-zheng commented:
@rxwei I see - I'll disable `self` differentiation parameter for `#gradient`, then proceed with this PR.

dan-zheng committed:
[WIP] #gradient and #valueAndGradient no longer support 'self' parameter.

dan-zheng committed:
Merge branch 'tensorflow' into libsyntax-autodiff

dan-zheng commented:
`self` differentiation parameter has been disabled for `#gradient`.
libSyntax has been updated so that `#gradient` no longer shares param list syntax with `@differentiable`.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/Parse/ParseType.cpp:
```diff
@@ -581,18 +581,47 @@ Parser::parseTypeIdentifier(bool isParsingQualifiedDeclName) {
+    // SWIFT_ENABLE_TENSORFLOW
+    // Returns true if the last type identifier component has been parsed.
+    // Lambda is used only when `isParsingQualifiedDeclName` is true.
+    auto parsedLastComponent = [](Parser &P) {
```
Why not just capture self?

dan-zheng replied:
Done.

rxwei commented on lib/Parse/ParseType.cpp:
```diff
@@ -581,18 +581,47 @@ Parser::parseTypeIdentifier(bool isParsingQualifiedDeclName) {
+
+    // SWIFT_ENABLE_TENSORFLOW
+    // Returns true if the last type identifier component has been parsed.
+    // Lambda is used only when `isParsingQualifiedDeclName` is true.
```
Where’s Lambda? Don’t refer to this function as “lambda”, just call it “this function” :)

dan-zheng replied:
Done in https://github.com/apple/swift/pull/17725/commits/fbbd8dcb5cad897f5fc01436e76ea4ecf11338ef.

dan-zheng committed:
Minor cleanup.

Address comments by @rxwei.

rxwei commented on utils/gyb_syntax_support/ExprNodes.py:
```diff
@@ -570,6 +570,82 @@
+    # SWIFT_ENABLE_TENSORFLOW
+    # Expression generalizing #gradient and #valueAndGradient.
+    # e.g. #gradient(foo(_:_:), wrt: .0, .1)
+    Node('GradientExpr', kind='Expr',
```
Note that this name is inconsistent with the base class name for `GradientExpr` and `ValueAndGradientExpr`: `ReverseAutoDiffExpr`. It needs to be changed to `ReverseAutoDiffExpr` to avoid confusion.
